src/vcSHP.cpp

```diff
@@ -0,0 +1,741 @@
+{
+  udResult result;
+
+  udFile *pFile = nullptr;
+  uint8_t *headerBuffer = nullptr;
+  uint32_t offset = 0;
+
+  vcSHP_Header &header = pSHP->shpHeader;
+  UD_ERROR_CHECK(udFile_Open(&pFile, pFilename, udFOF_Read));
+
+  headerBuffer = udAllocType(uint8_t, 100, udAF_Zero);
+  UD_ERROR_CHECK(udFile_Read(pFile, headerBuffer, 100 * sizeof(uint8_t)));
+  // 00 00 27 0a
+  UD_ERROR_IF((headerBuffer[0] != 0 || headerBuffer[1] != 0 || headerBuffer[2] != 0x27 || headerBuffer[3] != 0x0a), udR_ParseError);
+
+  header.fileLength = (headerBuffer[24] << 24) | (headerBuffer[25] << 16) | (headerBuffer[26] << 8) | headerBuffer[27];
+  if (header.fileLength < UINT_MAX / 2)
+    header.fileLength *= 2;
+  else
+    header.fileLength = (UINT_MAX / 2) * 2;
+
+  header.version = (headerBuffer[31] << 24) | (headerBuffer[30] << 16) | (headerBuffer[29] << 8) | headerBuffer[28];
+  header.shapeType = (vcSHP_Type)((int32_t)((headerBuffer[35] << 24) | (headerBuffer[34] << 16) | (headerBuffer[33] << 8) | headerBuffer[32]));
+
+  offset = 36;
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MBRmin.x, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MBRmin.y, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MBRmax.x, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MBRmax.y, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.ZRange.x, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.ZRange.y, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MRange.x, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MRange.y, headerBuffer);
+
+  UD_ERROR_IF(offset != 100, udR_ParseError);
+  udFree(headerBuffer);
+
+  UD_ERROR_CHECK(pSHP->shpRecords.Init(32));
+  while (offset < header.fileLength)
+  {
+    result = vcSHP_LoadShpRecord(pSHP, pFile, offset, header.shapeType);
+    UD_ERROR_CHECK(result);
+  }
+
+epilogue:
+  udFile_Close(&pFile);
+
+  if (headerBuffer)
+    udFree(headerBuffer);
+
+  return result;
+}
+
+void vcSHP_LoadDbfFile(vcSHP *pSHP, const char *pFilename)
+{
+  vcDBF_Create(&pSHP->pDBF);
+  vcDBF_Load(&pSHP->pDBF, pFilename);
+}
+
+udResult vcSHP_LoadFileGroup(vcSHP *pSHP, const char *pFilename)
+{
+  if (pSHP == nullptr || pFilename == nullptr)
+    return udR_InvalidParameter_;
+
+  int i = 1;
+  if (*((uint8_t *)&i) == 1)
+    pSHP->localBigEndian = false;
+  else
+    pSHP->localBigEndian = true;
+
+  // load shp file
+  udResult result;
+  result = vcSHP_LoadShpFile(pSHP, pFilename);
+  if (result != udR_Success)
+  {
+    vcSHP_Release(pSHP);
+    return result;
+  }
+
+  printf("shape file load successfully %s\n", pFilename);
+
+  char extraFileName[256] = "";
+  int32_t nameLen = (int32_t)strlen(pFilename) - 3;
+  memcpy(extraFileName, pFilename, nameLen);
+  memcpy(extraFileName + nameLen, "dbf", 3);
```

Comment:
Consider using `udFilename` and using the `SetExtension` member function instead of this.

Reply:
fixed.